mqtt-client-C-paho/emqx_file_transfer.c

```diff
@@ -0,0 +1,290 @@
+            "  \"size\": %ld,\n"
+            "  \"expire_at\": %ld,\n"
+            "  \"segments_ttl\": %ld,\n"
+            "  \"user_data\": {}\n"
+            "}",
+            file_name,
+            file_size,
+            expire_time_s_since_epoch,
+            segments_ttl_seconds);
+    if (rc < 0 || rc >= buf_size) {
+        printf("Failed to create payload for initial message\n");
+        return -1;
+    }
+    // Create topic of the form $file/{file_id}/init for initial message
+    char topic[buf_size];
+    MQTTClient_deliveryToken token;
+    rc = snprintf(topic, buf_size, "$file/%s/init", file_id);
+    if (rc < 0 || rc >= buf_size) {
+        printf("Failed to create topic for initial message\n");
+        return -1;
+    }
+    // Publish initial message
+    if (DEBUG) {
+        printf("Publishing initial message to topic %s\n", topic);
+        printf("Payload: %s\n", payload);
+    }
+    rc = MQTTClient_publish(client, topic, strlen(payload), payload, 1, 0, &token);
+    if (rc != MQTTCLIENT_SUCCESS) {
+        printf("Failed to publish message, return code %d\n", rc);
+        return -1;
+    }
+    rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
+    if (rc != MQTTCLIENT_SUCCESS) {
+        printf("Failed to publish message, return code %d\n", rc);
+        return -1;
+    }
+    // Read binary chunks of max size 1024 bytes and publish them to the broker
+    // The chunks are published to the topic of the form $file/{file_id}/{offset}
+    // The chunks are read into the payload
+    size_t chunk_size = 1024;
+    size_t offset = 0;
+    size_t read_bytes;
+    while ((read_bytes = fread(payload, 1, chunk_size, fp)) > 0) {
+        rc = snprintf(topic, buf_size, "$file/%s/%lu", file_id, offset);
+        if (rc < 0 || rc >= buf_size) {
+            printf("Failed to create topic for file chunk\n");
+            return -1;
+        }
+        if (DEBUG) {
+            printf("Publishing file chunk to topic %s offset %lu\n", topic, offset);
+        }
+        rc = MQTTClient_publish(client, topic, read_bytes, payload, 1, 0, &token);
+        if (rc != MQTTCLIENT_SUCCESS) {
+            printf("Failed to publish file chunk, return code %d\n", rc);
+            return -1;
+        }
+        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
+        if (rc != MQTTCLIENT_SUCCESS) {
+            printf("Failed to publish file chunk, return code %d\n", rc);
+            return -1;
+        }
+        offset += read_bytes;
+    }
+    // Send final message to the topic $file/{file_id}/fin/{file_size} with an empty payload
+    rc = snprintf(topic, buf_size, "$file/%s/fin/%ld", file_id, file_size);
+    if (rc < 0 || rc >= buf_size) {
+        printf("Failed to create topic for final message\n");
+        return -1;
+    }
+    if (DEBUG) {
+        printf("Publishing final message to topic %s\n", topic);
+    }
+    rc = MQTTClient_publish(client, topic, 0, "", 1, 0, &token);
```

Comment:
Q: Why don't we follow with `MQTTClient_waitForCompletion` here?

Reply:
Fixed in: [88b1f72](https://github.com/emqx/MQTT-Client-Examples/pull/112/commits/88b1f72f5d0dac5495a82acc0916bccfe028fd02)